styx-client/src/main/java/com/spotify/styx/client/auth/GoogleIdTokenAuth.java

```diff
@@ -103,22 +114,85 @@ private String getDefaultGCEIdToken(String targetAudience) throws IOException {
         .parseAsString();
   }
 
-  private String getServiceAccountToken(GoogleCredential credential, String targetAudience)
+  private String getServiceAccountToken(ServiceAccountCredentials credential, String targetAudience)
       throws IOException, GeneralSecurityException {
-    log.debug("Fetching service account id token for {}", credential.getServiceAccountId());
+    log.debug("Fetching service account id token for {}", credential.getAccount());
     final TokenRequest request = new TokenRequest(
         this.httpTransport, JSON_FACTORY,
-        new GenericUrl(credential.getTokenServerEncodedUrl()),
+        new GenericUrl(credential.getTokenServerUri()),
         "urn:ietf:params:oauth:grant-type:jwt-bearer");
     final Header header = jwtHeader();
     final Payload payload = jwtPayload(
-        targetAudience, credential.getServiceAccountId(), credential.getTokenServerEncodedUrl());
+        targetAudience, credential.getAccount(), credential.getTokenServerUri().toString());
     request.put("assertion", JsonWebSignature.signUsingRsaSha256(
-        credential.getServiceAccountPrivateKey(), JSON_FACTORY, header, payload));
+        credential.getPrivateKey(), JSON_FACTORY, header, payload));
     final TokenResponse response = request.execute();
     return (String) response.get("id_token");
   }
 
+  private String getImpersonatedIdToken(ImpersonatedCredentials credentials, String targetAudience) throws IOException {
```

Comment:
We should add unit tests for these new kind of credentials

Also this class looks complicated with a lot of nontrivial logic smashed all together

Reply:
Yeah like I mentioned above, we should probably set up a GCP project where we can test this code. I do not believe that _unit_ tests would give much confidence here due to tight coupling with the IAM service api. We cannot emulate or mock IAM well enough to know whether it would work in production. We need to integration test this against the real IAM api.

Reply:
It makes sense